pkg/p2p/server.go

```diff
@@ -0,0 +1,872 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package p2p
+
+import (
+	"context"
+	"reflect"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/version"
+	"github.com/pingcap/ticdc/pkg/workerpool"
+	"github.com/pingcap/ticdc/proto/p2p"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	gRPCPeer "google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	waitUnregisterHandleTimeout = time.Second * 5
+)
+
+// MessageServerConfig stores configurations for the MessageServer
+type MessageServerConfig struct {
+	// The maximum number of entries to be cached for topics with no handler registered
+	MaxTopicPendingCount int
+	// The maximum number of unhandled internal tasks for the main thread.
+	MaxPendingTaskCount int
+	// The size of the channel for pending messages before sending them to gRPC.
+	SendChannelSize int
+	// The interval between ACKs.
+	AckInterval time.Duration
+	// The size of the goroutine pool for running the handlers.
+	WorkerPoolSize int
+}
+
+// cdcPeer is used to store information on one connected client.
+type cdcPeer struct {
+	PeerID string // unique ID of the client
+	Epoch  int64  // epoch is increased when the client retries.
+
+	// necessary information on the stream.
+	sender *streamHandle
+
+	metricsAckCount prometheus.Counter
+}
+
+func newCDCPeer(senderID NodeID, epoch int64, sender *streamHandle) *cdcPeer {
+	return &cdcPeer{
+		PeerID: senderID,
+		Epoch:  epoch,
+		sender: sender,
+		metricsAckCount: serverAckCount.With(prometheus.Labels{
+			"to": senderID,
+		}),
+	}
+}
+
+func (p *cdcPeer) abortWithError(ctx context.Context, err error) {
+	if err1 := p.sender.Send(ctx, errorToRPCResponse(err)); err1 != nil {
+		log.Warn("could not send error to peer", zap.Error(err),
+			zap.NamedError("send-err", err1))
+		return
+	}
+	log.Debug("sent error to peer", zap.Error(err))
```

Comment:
What about `Warn` or `Info`?

Reply:
`deregisterPeer` already prints the `err`. This debug log here is for debugging.